### PR TITLE
[RHCLOUD-45039] Unify empty filter param handling across V2 list endpoints

### PR DESCRIPTION
**Pull request by dsova06**
## Link(s) to Jira
- [RHCLOUD-45039](https://issues.redhat.com/browse/RHCLOUD-45039)

## Description of Intent of Change(s)

**Problem:** V2 list endpoints handle empty query parameters (`?name=`, `?resource_type=`) inconsistently:
- Roles V2 list: treats empty as unset (skip filter) -- this is the reference pattern
- RoleBinding list: returns 400 on empty string
- RoleBinding by-subject: mixed behavior (some fields allow blank, others don't)
- Workspace list: correct behavior but uses raw view-level parsing instead of the serializer+service pattern

**Decision (confirmed with Sneha and Jay):** Empty string at field level = treat as unset (skip the filter) for all optional filter params. Conditionally required params continue to enforce requirements in cross-field `validate()`.

### What changed

**Role binding serializers:**
- Added `allow_blank=True` + `validate_<field>` normalizers to all optional filter fields in `RoleBindingListInputSerializer`, `RoleBindingInputSerializer` (by-subject), and `UpdateRoleBindingRequestSerializer`
- Converted `role_id` and `subject_id` from `UUIDField` to `CharField` with manual UUID validation (consistent with `RoleV2ListSerializer.validate_resource_id` pattern)
- Removed redundant `.strip()` checks in `_validate_resource_identifiers` that would crash on `None` after normalization

**Workspace list refactor:**
- Created `WorkspaceListInputSerializer` with empty-string normalization, NUL byte rejection, comma-separated type parsing, UUID validation for `parent_id`/`ids`, and cross-field defaulting (ids without explicit type defaults to standard)
- Added `WorkspaceService.list()` method for domain filtering
- Simplified `WorkspaceViewSet.list()` from ~60 lines of inline parsing to a clean serializer+service call

### Breaking change

RoleBinding list (`GET /api/rbac/v2/role_bindings/`) will stop returning 400 for empty filter params like `?role_id=` or `?resource_type=`. Empty params are now treated as unset (no filter applied), matching the Roles V2 behavior.

## Local Testing

```bash
# Run targeted tests
pipenv run tox -e py312-fast -- tests.management.role_binding.test_serializer tests.management.workspace.test_serializer tests.management.workspace.test_view

# Run full suite
DJANGO_LOG_LEVEL=CRITICAL RBAC_LOG_LEVEL=CRITICAL pipenv run tox -e py312-fast
```

All 3491 tests pass, lint passes.

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [x] are there any pre/post merge actions required?
- [x] are these changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [x] does this require migration changes? No
- [ ] is there known, direct impact to dependent teams/components?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

[RHCLOUD-45039]: https://redhat.atlassian.net/browse/RHCLOUD-45039?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Unify handling of empty string query parameters for V2 role binding and workspace list endpoints so that empty optional filters are treated as unset, and refactor workspace list filtering into a dedicated serializer and service layer.

New Features:
- Introduce WorkspaceListInputSerializer to validate and normalize workspace list query parameters, including type, name, parent_id, and ids.
- Add WorkspaceService.list() to encapsulate domain-level filtering for workspace queries.

Enhancements:
- Normalize optional role binding filter parameters across list, by-subject, and update serializers so empty strings are treated as unset while preserving cross-field validation rules.
- Refactor the workspace list view to delegate query parameter parsing and filtering to the new serializer and WorkspaceService method, simplifying the view logic.

Tests:
- Add comprehensive serializer tests for WorkspaceListInputSerializer covering type parsing, UUID validation, cross-field defaults, and NUL-byte rejection.
- Extend role binding serializer tests to cover empty-as-unset normalization and related cross-field validation behavior across list, by-subject, and update serializers.

### CODE REVIEW

**Comment by dsova06** [comment]
@sourcery-ai review

**Review by dsova06 — COMMENTED** [review]
Self-review: clean, well-organized PR. The empty-as-unset normalization pattern is consistent across all three serializers, the workspace refactor to serializer+service matches the established V2 pattern, and test coverage is thorough. Two minor notes:

### Would be good to clean up

1. **UpdateRoleBindingRequestSerializer test gap** -- `RoleBindingListInputSerializer` and `RoleBindingInputSerializer` both have explicit tests for empty resource_id/resource_type/resource_tenant_org_id normalization. `UpdateRoleBindingRequestSerializerTests` doesn't, even though it now uses the same `staticmethod(_normalize_blank_or_none)` pattern. Adding a couple tests there would complete the coverage symmetry (empty resource_id triggers cross-field error since subject_id/subject_type are required, empty resource_tenant_org_id normalizes to None).

### Something to think about

1. **NUL byte scan scope in WorkspaceListInputSerializer** -- `to_internal_value` iterates `data.items()` which is the full `request.query_params` QueryDict, so it checks NUL bytes in ALL query params (including pagination `limit`/`offset`, `order_by`, etc.), not just the 4 declared filter fields. The old `clean_query_param` only checked specific params. This is stricter, and NUL bytes are never legitimate, so it's fine in practice -- just worth being aware of the wider blast radius.

2. **Whitespace handling asymmetry between helpers** -- `_normalize_uuid_or_none` handles whitespace-only strings explicitly (`if not value or not value.strip()`), while `_normalize_blank_or_none` relies on DRF's `trim_whitespace=True` default to strip whitespace before the validator runs (`return value or None`). Both work correctly in practice since all usages are on CharField, but if `_normalize_blank_or_none` were ever called outside a CharField context, whitespace-only strings would pass through.

Nice use of `staticmethod()` to eliminate the boilerplate -- reads well and makes the pattern obvious at a glance.

File: tests/management/role_binding/test_serializer.py [hunk 1]
```diff
@@ -1005,6 +1018,59 @@ def test_nul_bytes_stripped_from_subject_id(self):
...
+    def test_all_empty_filter_params_valid(self):
+        """Test that all filter params as empty strings produce valid result with None values."""
+        s = RoleBindingListInputSerializer(
+            data={
+                "role_id": "",
+                "resource_id": "",
+                "resource_type": "",
+                "subject_type": "",
+                "subject_id": "",
+                "granted_subject_type": "",
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/role_binding/test_serializer.py:1032]: Fixed: added fields and order_by to the assertion loop in test_all_empty_filter_params_valid. Both normalize to None for empty strings.

**Comment by dsova06** [comment]
@sourcery-ai resolve

**Comment by dsova06** [comment]
Re: Sourcery second review feedback:

**_normalize_blank_or_none should strip()**: DRF's CharField has trim_whitespace=True by default, which strips whitespace before the validator runs. All usages of this helper are on CharField fields, so whitespace-only strings are already handled. Adding explicit .strip() would be defense-in-depth but unnecessary given the current usage.

**Cross-module UUID validation sharing**: The workspace and role-binding UUID validators have different semantics -- workspace returns strings (for queryset filtering), role-binding returns uuid.UUID objects (for service layer compatibility). Different error message formats too. Extracting a shared helper would add coupling between modules for marginal dedup.

**Comment by lpichler** [comment]
## Code Review Findings

### 1. Minor: `exclude_sources` ChoiceField not handled for empty string (LOW)

**File**: `rbac/management/role_binding/serializer.py`
**Lines**: `RoleBindingListInputSerializer.exclude_sources` field definition

The `exclude_sources` ChoiceField does not have `allow_blank=True`. If a client sends `?exclude_sources=`, it will receive a 400 validation error because empty string is not in the choices list. While this is arguably correct for an enum field with a default value (it differs from free-text filters), it is inconsistent with the PR's stated goal of treating empty optional params as unset. The inconsistency could surprise API consumers who test empty params uniformly.

**Recommendation**: Consider adding `allow_blank=True` with a `validate_exclude_sources` normalizer that maps blank to the default `ExcludeSources.NONE`, or document this as an intentional exception.

**Severity**: Low -- `exclude_sources` is an enum with a sensible default, so rejecting empty is defensible.

**Review by lpichler — APPROVED** [review]

**Comment by dsova06** [comment]
@lpichler Fixed: added `allow_blank=True` to `exclude_sources` ChoiceField with a `validate_exclude_sources` normalizer that maps blank to the default `ExcludeSources.NONE`. Updated `test_all_empty_filter_params_valid` to include `exclude_sources=""` and assert it defaults correctly. Thanks for catching the inconsistency.
